include/openssl/ssl.h

```diff
@@ -1463,8 +1463,20 @@ __owur BIO *BIO_new_buffer_ssl_connect(SSL_CTX *ctx);
 __owur int BIO_ssl_copy_session_id(BIO *to, BIO *from);
 void BIO_ssl_shutdown(BIO *ssl_bio);
 
-__owur int SSL_CTX_set_cipher_list(SSL_CTX *, const char *str);
 __owur SSL_CTX *SSL_CTX_new(const SSL_METHOD *meth);
+/* As an option for apps, TLSv1.3 can be disabled */
+#ifndef OPENSSL_DISABLE_TLS1_3
+/*
+ * With TLSv1.3 enabled, we make sure that apps linked against the shared
+ * libraries of earlier OpenSSL versions don't get it by mistake, as TLSv1.3
+ * is different enough to be able to cause faults for the unknowing.
```

Comment:
Is this actually true? Is common 1.1.0 application **bound** to fail if library negotiates TLS 1.3? At least 1.1.0 s_client and s_server don't. And if they don't, then why would we want to **deny** them the option to use the protocol? I'd say opposite is more natural. Indeed, they concentrate on application logic and we concentrate on transport mechanism. If anything, we should facilitate use on new protocol with old applications.

Reply:
I've been trying to ask @mattcaswell if there's a subset of functionality that still works, and I keep getting the answer back that TLSv1.3 works differently.  I take that as there not being any guarantee, and that it's pure luck that `s_client` and `s_server` work when no extra options are added (have you tried with different options?  How long before it fails surprisingly?)

One of the test programs that I've been worried about is `test/sslapitest`.  After all, it *tests the behaviour of the SSL API!*  I'm thinking that if that one fails (i.e. that one from 1.1.0 linked with 1.1.1 libraries), we do have a regression problem.

Reply:
This is, btw, a discussion that could happen in #5944

Reply:
Ahm, do you mean when a program is just re-compiled then it will fail in a surprising way?

Reply:
> I keep getting the answer back that TLSv1.3 works differently.

It does work differently. That doesn't mean that all applications will start failing. Quite the opposite. I expect most applications to work without noticing TLSv1.3. However, there are certain scenarios where unexpected things might start happening. Some examples:

- if you expect your session to have been created during the handshake then you're going to be disappointed (it happens post-handshake). This means some callbacks could happen when an application isn't expecting them.
- If you're expecting a particular ciphersuite to be negotiated (because you disabled others) then that might not happen because TLSv1.3 ciphersuite configuration happens differently. This could actually cause an application to fail in some circumstances.
- If your application is using custom extensions you might suddenly find them not being used when they always were before because they are not compatible with TLSv1.3
- etc

There's probably a whole load of other things. Most of these problems are edge cases. What we don't know is what kind of proportion of applications will fall foul of them.

Reply:
I'd like to repeat the suggestion that libssl could look at certain facts (sigalgs derived from server certs, ciphersuites, whatever else that's *possible* to check) and adapt the range of TLS versions to be negotiated.  For example, is `s_server` is started with a DSA server certificate, TLSv1.2 would be the newest version negotiated, rather than selecting TLSv1.3 and then generating an error because the sigalg doesn't fit...